ResourceWarning: unclosed event loop

using pytest-asyncio asyncio 0.23.2

```python
import pytest

import pytest_asyncio


@pytest_asyncio.fixture()
async def demo():
    yield


@pytest.mark.asyncio
async def test_aiohttp_test_client_json(demo):
    pass


def test_redirect():

    import asyncio
    async def amain():
        pass
    asyncio.run(amain())

    import gc
    gc.collect()
```

results in

```
REQUESTS_CA_BUNDLE=/home/graingert/projects/vcrpy/.tox/py311-aiohttp/lib/python3.11/site-packages/pytest_httpbin/certs/cacert.pem ./.tox/py311-aiohttp/bin/pytest tests/integration/test_aiohttp.py -s
================================================================================ test session starts =================================================================================
platform linux -- Python 3.11.7+, pytest-7.4.3, pluggy-1.3.0
rootdir: /home/graingert/projects/vcrpy
configfile: pyproject.toml
plugins: aiohttp-1.0.5, httpbin-2.0.0, cov-4.1.0, asyncio-0.23.2
asyncio: mode=Mode.STRICT
collected 2 items                                                                                                                                                                    

tests/integration/test_aiohttp.py ..

================================================================================== warnings summary ==================================================================================
tests/integration/test_aiohttp.py::test_redirect
  /usr/lib/python3.11/asyncio/base_events.py:692: ResourceWarning: unclosed event loop <_UnixSelectorEventLoop running=False closed=False debug=False>
    _warn(f"unclosed event loop {self!r}", ResourceWarning, source=self)
  Enable tracemalloc to get traceback where the object was allocated.
  See https://docs.pytest.org/en/stable/how-to/capture-warnings.html#resource-warnings for more info.

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
============================================================================ 2 passed, 1 warning in 0.02s ===========================================================================
```

I have the same issue in versions 0.22.0+, including 0.23.0a0.

I first spent a ton of time trying to hunt down a new `ResourceWarning` I seemingly randomly started getting in my tests (maybe due to the more tests I added and thus more time to fail with the resource warning, unsure):

```
>               warnings.warn(pytest.PytestUnraisableExceptionWarning(msg))
E               pytest.PytestUnraisableExceptionWarning: Exception ignored in: <socket.socket fd=-1, family=1, type=1, proto=0>
E               
E               Traceback (most recent call last):
E                 File "/Users/phillip/.pyenv/versions/3.11.6/lib/python3.11/json/decoder.py", line 353, in raw_decode
E                   obj, end = self.scan_once(s, idx)
E                              ^^^^^^^^^^^^^^^^^^^^^^
E               ResourceWarning: unclosed <socket.socket fd=13, family=1, type=1, proto=0>

```

I couldn't for the life of me figure out where I had an unclosed socket. So I finally admitted defeat and silenced the `ResourceWarning`, only to have this next error happen:

```
>               warnings.warn(pytest.PytestUnraisableExceptionWarning(msg))
E               pytest.PytestUnraisableExceptionWarning: Exception ignored in: <function BaseEventLoop.__del__ at 0x104bed760>
E               
E               Traceback (most recent call last):
E                 File "/Users/phillip/.pyenv/versions/3.11.6/lib/python3.11/asyncio/base_events.py", line 692, in __del__
E                   _warn(f"unclosed event loop {self!r}", ResourceWarning, source=self)
E               ResourceWarning: unclosed event loop <_UnixSelectorEventLoop running=False closed=False debug=False>
```

The results seemed to change depending on how many tests I was running in my test suite.

Downgrading to 0.21.0 fixed both problems.

--- a/changelog.d/724.fixed.rst
+++ b/changelog.d/724.fixed.rst
@@ -1,0 +1,1 @@
+Fixed a ``ResourceWarning: unclosed event loop`` warning that could occur when a synchronous test called ``asyncio.run()`` or otherwise unset the current event loop after pytest-asyncio had run an async test or fixture.

--- a/pytest_asyncio/plugin.py
+++ b/pytest_asyncio/plugin.py
@@ -806,23 +806,13 @@ def _temporary_event_loop(loop: AbstractEventLoop) -> Iterator[None]:
 @contextlib.contextmanager
 def _temporary_event_loop_policy(
     policy: AbstractEventLoopPolicy,
-    *,
-    has_custom_factory: bool,
 ) -> Iterator[None]:
     old_loop_policy = _get_event_loop_policy()
-    if has_custom_factory:
-        old_loop = None
-    else:
-        try:
-            old_loop = _get_event_loop_no_warn()
-        except RuntimeError:
-            old_loop = None
     _set_event_loop_policy(policy)
     try:
         yield
     finally:
         _set_event_loop_policy(old_loop_policy)
-        _set_event_loop(old_loop)
 
 
 def _get_event_loop_policy() -> AbstractEventLoopPolicy:
@@ -1042,10 +1032,7 @@ def _create_scoped_runner_fixture(scope: _ScopeName) -> Callable:
     ) -> Iterator[Runner]:
         new_loop_policy = event_loop_policy
         debug_mode = _get_asyncio_debug(request.config)
-        with _temporary_event_loop_policy(
-            new_loop_policy,
-            has_custom_factory=_asyncio_loop_factory is not None,
-        ):
+        with _temporary_event_loop_policy(new_loop_policy):
             runner = Runner(
                 debug=debug_mode,
                 loop_factory=_asyncio_loop_factory,
@@ -1068,6 +1055,9 @@ def _create_scoped_runner_fixture(scope: _ScopeName) -> Callable:
                             _RUNNER_TEARDOWN_WARNING % traceback.format_exc(),
                             RuntimeWarning,
                         )
+            finally:
+                if _asyncio_loop_factory is not None:
+                    _set_event_loop(None)
 
     return _scoped_runner
 

--- a/tests/test_set_event_loop.py
+++ b/tests/test_set_event_loop.py
@@ -99,6 +99,40 @@ def test_set_event_loop_none_class(pytester: Pytester, loop_breaking_action: str
     result.assert_outcomes(passed=3)
 
 
+def test_asyncio_run_after_async_fixture_does_not_leak_loop(
+    pytester: Pytester,
+):
+    pytester.makeini(dedent("""\
+            [pytest]
+            asyncio_default_fixture_loop_scope = function
+            """))
+    pytester.makepyfile(dedent("""\
+            import asyncio
+            import gc
+            import pytest
+            import pytest_asyncio
+
+            pytest_plugins = "pytest_asyncio"
+
+            @pytest_asyncio.fixture
+            async def async_fixture():
+                yield
+
+            @pytest.mark.asyncio
+            async def test_async_function_uses_async_fixture(async_fixture):
+                pass
+
+            def test_collect_unclosed_loops():
+                async def amain():
+                    pass
+
+                asyncio.run(amain())
+                gc.collect()
+            """))
+    result = pytester.runpytest_subprocess("-W", "error")
+    result.assert_outcomes(passed=2)
+
+
 @pytest.mark.parametrize("test_loop_scope", ("module", "package", "session"))
 @pytest.mark.parametrize(
     "loop_breaking_action",

